joblib/parallel.py

```diff
@@ -46,21 +46,58 @@
 # managed by ``parallel_backend``.
 DEFAULT_BACKEND = 'loky'
 DEFAULT_N_JOBS = 1
+DEFAULT_THREAD_BACKEND = 'threading'
 
 # Thread local value that can be overridden by the ``parallel_backend`` context
 # manager
 _backend = threading.local()
 
+VALID_BACKEND_HINTS = ('processes', 'threads', None)
+VALID_BACKEND_CONSTRAINTS = ('sharedmem', None)
 
-def get_active_backend():
+
+def get_active_backend(prefer=None, require=None, verbose=0):
     """Return the active default backend"""
-    active_backend_and_jobs = getattr(_backend, 'backend_and_jobs', None)
-    if active_backend_and_jobs is not None:
-        return active_backend_and_jobs
+    if prefer not in VALID_BACKEND_HINTS:
+        raise ValueError("prefer=%r is not a valid backend hint, "
+                         "expected one of %r" % (prefer, VALID_BACKEND_HINTS))
+    if require not in VALID_BACKEND_CONSTRAINTS:
+        raise ValueError("require=%r is not a valid backend constraint, "
+                         "expected one of %r"
+                         % (require, VALID_BACKEND_CONSTRAINTS))
+
+    if prefer == 'processes' and require == 'sharedmem':
+        raise ValueError("prefer == 'processes' and require == 'sharedmem'"
+                         " are inconsistent settings")
+    backend_and_jobs = getattr(_backend, 'backend_and_jobs', None)
+    if backend_and_jobs is not None:
+        # Try to use the backend set by the user with the context manager.
+        backend, n_jobs = backend_and_jobs
+        supports_sharedmem = getattr(backend, 'supports_sharedmem', False)
+        if require == 'sharedmem' and not supports_sharedmem:
+            # This backend does not match the shared memory constraint:
+            # fallback to the default thead-based backend.
+            sharedmem_backend = BACKENDS[DEFAULT_THREAD_BACKEND]()
+            if verbose >= 10:
+                print("Using %s as joblib.Parallel backend instead of %s "
```

Comment:
This might be a project idiom, but I'd prefer a warning in this case rather than print if `verbose >= 10`. Warnings can be silenced as needed, but if I wrote code to use a certain backend and that backend is being ignored I'd like to know even if I have `verbose` set to 0.

Reply:
Actually, I don't want to issue a warning in this case because the user cannot do anything to "fix" the problem: take the example fo random forests in scikit-learn: the current implementation relies on shared memory semantics in the prediction loop but not in the main fit loop. If the users is calling RF in a nested cross-val loop (which can itself benefit from parallelism). If the user uses a context manager to use the dask-distributed backend to parallelize as much as possible on all the levels, the prediction part should stay sequential (the default behavior) while all other parallel calls will benefit from dask (which would be more than enough to saturate all the cores in typical uses cases). There is no point in issuing a noisy warning that the user will not understand without understanding the details of the inner code of scikit-learn.

I have changed the verbosity of the parallel call itself to make it explicit which backend is used for each individual call however.